### PR TITLE
Save the registered factories and look up from this collection first

### PR DESCRIPTION
**Pull request by Haoning-Sun**
Fixes https://github.com/Alluxio/alluxio/issues/16001
### What changes are proposed in this pull request?

Save the registered factories and look up from this collection first.

### Why are the changes needed?
Reduce the memory consumption of metaspace and avoid master crash due to OutOfMemory of metaspace.

### Does this PR introduce any user facing changes?
No.

### CODE REVIEW

**Comment by LuQQiu** [comment]
@LuQQiu

**Comment by LuQQiu** [comment]
@yuyang733 Please help take a look

**Comment by LuQQiu** [comment]
https://github.com/Alluxio/alluxio/issues/16001

**Review by jiacheliu3 — APPROVED** [review]
The caching approach LGTM. Thanks for the fix!

Just curious @Haoning-Sun , could you provide more details about the metaspace OOM you see? For example with how many jars under `lib/`, and how much memory was metaspace consuming?

@maobaolong  One side comment is this change may not work with https://github.com/Alluxio/alluxio/pull/16018/files because the strict version match takes place after filtering on the cache, and if you add a jar into `lib/` after Alluxio is started, the new jar will not be picked up. So if you mount with a strict version expecting the new jar, that new jar will not be loaded and your mount will fail. You will then have to restart the Alluxio master.

File: core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java [hunk 1]
```diff
@@ -200,6 +220,7 @@ private void scan(List<File> files, List<T> factories) {
           LOG.debug("Discovered a factory implementation {} - {} in jar {}", factory.getClass(),
               factory, jarPath);
           register(factory, factories);
+          register(factory);
```
[the remark this answers is not in the document]
Comment (Haoning-Sun) [reply core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java:223]: This method is called to add factory to mFactories, while the above method only adds it to local variables.

**Comment by Haoning-Sun** [comment]
The following is the heap info before and after mounting an ozone path.

- command
```
bin/alluxio fs mount --option alluxio-underfs.hdfs.configuration=ozone-site.xml --option alluxio.underfs.version=1.2.1 /ozone2 o3fs://jan2.hive
```

- before mounting

```
16533:
 PSYoungGen      total 557568K, used 90312K [0x0000000715580000, 0x0000000742e80000, 
0x00000007c0000000)
  eden space 485376K, 18% used [0x0000000715580000,0x000000071adb2110,0x0000000732f80000)
  from space 72192K, 0% used [0x0000000739c00000,0x0000000739c00000,0x000000073e280000)
  to   space 111104K, 0% used [0x0000000732f80000,0x0000000732f80000,0x0000000739c00000)
 ParOldGen       total 305152K, used 176268K [0x00000005c0000000, 0x00000005d2a00000, 0x0000000715580000)
  object space 305152K, 57% used [0x00000005c0000000,0x00000005cac230b8,0x00000005d2a00000)
 Metaspace       used 102112K, capacity 105818K, committed 106624K, reserved 1142784K
  class space    used 12151K, capacity 12747K, committed 12928K, reserved 1048576K
```

- after mounting
```
16533:
 PSYoungGen      total 557568K, used 236427K [0x0000000715580000, 0x0000000742e80000, 0x00000007c0000000)
  eden space 485376K, 48% used [0x0000000715580000,0x0000000723c62c70,0x0000000732f80000)
  from space 72192K, 0% used [0x0000000739c00000,0x0000000739c00000,0x000000073e280000)
  to   space 111104K, 0% used [0x0000000732f80000,0x0000000732f80000,0x0000000739c00000)
 ParOldGen       total 305152K, used 176268K [0x00000005c0000000, 0x00000005d2a00000, 0x0000000715580000)
  object space 305152K, 57% used [0x00000005c0000000,0x00000005cac230b8,0x00000005d2a00000)
 Metaspace       used 126540K, capacity 130502K, committed 131072K, reserved 1165312K
  class space    used 14878K, capacity 15525K, committed 15616K, reserved 1048576K
```

It can be seen that the memory occupied by metaspace is growing.

**Review by Jackson-Wang-7 — APPROVED** [review]
I have discussed this change with you before, and I think the implementation LGTM. Thanks for that.

File: core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java:223]: I thought one method is actually calling the other?
https://github.com/Alluxio/alluxio/blob/a4a85ee0bf2b490d49d32e8c52d7968676bea7f2/core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java#L223

[the remark this answers is not in the document]
Comment (Jackson-Wang-7) [reply core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java:223]: line 223 is to register `factory` into temporary parameter `factories`
line 223 is to register `factory` into the global parameter `mFactories`

[the remark this answers is not in the document]
Comment (jiacheliu3) [reply core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java:223]: wow this is so misleading. @Haoning-Sun could you add a comment like below:
```
// Also add to the cache
register(factory);
```

[the remark this answers is not in the document]
Comment (Haoning-Sun) [reply core/common/src/main/java/alluxio/extensions/ExtensionFactoryRegistry.java:224]: Done.

**Comment by jiacheliu3** [comment]
@Haoning-Sun there are some UT failing not because of your code change and they have been fixed in the latest `master` https://github.com/Alluxio/alluxio/commit/446a255d348485c2d5378042d1d87753a1a3fc24. Could you rebase your branch and push again? I will merge after your rebase. Thanks and sorry about the trouble.

**Comment by Haoning-Sun** [comment]
@jiacheliu3 Rebase completed.

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by maobaolong** [comment]
@Haoning-Sun @jiacheliu3 @LuQQiu Thanks for submit and reviewed this PR, but I'm not quite understand how this PR can resolve my issue #16001

**Pull request merged**